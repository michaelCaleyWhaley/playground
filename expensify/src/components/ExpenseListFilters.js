import React, { Component } from "react";
import { connect } from "react-redux";
import { DateRangePicker } from "react-dates";
import {
  setTextFilter,
  sortByDate,
  sortByAmount,
  setStartDate,
  setEndDate,
} from "../actions/filters";

export class ExpenseListFilters extends Component {
  constructor() {
    super();
    this.state = {
      calendarFocused: null,
    };
  }

  onDatesChange = ({ startDate, endDate }) => {
    const { setStartDate, setEndDate } = this.props;
    setStartDate(startDate);
    setEndDate(endDate);
  };

  onFocusChange = (calendarFocused) => {
    this.setState({ calendarFocused });
  };

  onTextChange = (e) => {
    const { setTextFilter } = this.props;
    setTextFilter(e.target.value);
  };

  onSortChange = (e) => {
    const { sortByDate, sortByAmount } = this.props;
    if (e.target.value === "date") {
      sortByDate();
    } else if (e.target.value === "amount") {
      sortByAmount();
    }
  };

  render() {
    const { text, sortBy, startDate, endDate } = this.props.filters;
    return (
      <div>
        <input type="text" value={text} onChange={this.onTextChange} />
        <select value={sortBy} onChange={this.onSortChange}>
          <option value="date">Date</option>
          <option value="amount">Amount</option>
        </select>

        <DateRangePicker
          startDate={startDate}
          startDateId="expensesStartDate"
          endDate={endDate}
          endDateId="expensesEndDate"
          onDatesChange={this.onDatesChange}
          focusedInput={this.state.calendarFocused}
          onFocusChange={this.onFocusChange}
          numberOfMonths={1}
          isOutsideRange={() => false}
          showClearDates={true}
        />
      </div>
    );
  }
}

const mapStateToProps = (state) => ({
  filters: state.filters,
});

const mapDispatchToProps = (dispatch) => ({
  setTextFilter: (textValue) => dispatch(setTextFilter(textValue)),
  sortByDate: () => dispatch(sortByDate()),
  sortByAmount: () => dispatch(sortByAmount()),
  setStartDate: (startDate) => dispatch(setStartDate(startDate)),
  setEndDate: () => (endDate) => dispatch(setEndDate(endDate)),
});

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters);
