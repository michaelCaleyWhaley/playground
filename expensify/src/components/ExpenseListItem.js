import React from "react";
import { Link } from "react-router-dom";

const ExpenseListItem = ({ description, amount, createdAt, note, id }) => (
  <li>
    <Link to={`/edit/${id}`}>
      <h3>Description: {description}</h3>
    </Link>
    <p>Cost: Â£{amount}</p>
    <p>CreatedAt: {createdAt}</p>
    <p>Note: {note}</p>
  </li>
);

export default ExpenseListItem;
