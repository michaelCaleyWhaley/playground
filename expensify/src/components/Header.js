import React from "react";
import { NavLink } from "react-router-dom";

const Header = () => (
  <header>
    <h1>Expensify App</h1>
    <ul>
      <li>
        <NavLink activeClassName="is-active" exact={true} to="/">
          Dashboard
        </NavLink>
      </li>
      <li>
        <NavLink activeClassName="is-active" exact={true} to="/create">
          Create Expense
        </NavLink>
      </li>
      <li>
        <NavLink activeClassName="is-active" exact={true} to="/help">
          Help
        </NavLink>
      </li>
    </ul>
  </header>
);

export default Header;
