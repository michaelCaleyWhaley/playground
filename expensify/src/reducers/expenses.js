// expenses reducer
// default state
const expensesReducerDefaultState = [];
const expensesReducer = (state = expensesReducerDefaultState, action) => {
  switch (action.type) {
    case "ADD_EXPENSE":
      return [...state, action.expense];
    case "REMOVE_EXPENSE":
      return state.filter(expense => expense.id !== action.id);
      return state;
    case "EDIT_EXPENSE":
      return state.map(currentValue => {
        if (currentValue.id === action.id) {
          return {
            ...currentValue,
            ...action.updates,
          };
        } else {
          return currentValue;
        }
      });
    default:
      return state;
  }
};

export default expensesReducer;
