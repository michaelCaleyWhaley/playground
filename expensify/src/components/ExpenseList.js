import React from "react";
import { connect } from "react-redux";
import ExpenseListItem from "./ExpenseListItem";
import SelectExpenses from "../selectors/expenses";

export const ExpenseList = ({ expenses, filters }) => (
  <div>
    <h1>Expense List</h1>
    {filters.text && <p>Filter: {filters.text}</p>}
    <ul>
      {expenses.length === 0 ? (
        <p>No expenses</p>
      ) : (
        expenses.map((expense, index) => (
          <ExpenseListItem key={index} {...expense} />
        ))
      )}
    </ul>
  </div>
);

const mapStateToProps = (state) => ({
  ...state,
  expenses: SelectExpenses(state.expenses, state.filters),
});
export default connect(mapStateToProps)(ExpenseList);
