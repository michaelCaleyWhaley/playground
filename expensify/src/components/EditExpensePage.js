import React from "react";
import { connect } from "react-redux";
import ExpenseForm from "./ExpenseForm";
import { editExpense, removeExpense } from "../actions/expenses";

export class EditExpensePage extends React.Component {
  onSubmit = (expense) => {
    const { props } = this;
    props.editExpense(props.expense.id, expense);
    props.history.push("/");
  };

  onClick = () => {
    const { props } = this;
    props.removeExpense(props.expense.id);
    props.history.push("/");
  };

  render() {
    const { props } = this;

    return (
      <div>
        <ExpenseForm expense={props.expense} onSubmit={this.onSubmit} />
        <button onClick={this.onClick}>Remove</button>
      </div>
    );
  }
}

const mapDispatchToProps = (dispatch) => ({
  removeExpense: (expenseId) => {
    dispatch(removeExpense({ id: expenseId }));
  },
  editExpense: (expenseId, expense) => {
    dispatch(editExpense(expenseId, expense));
  },
});

const mapStateToProps = (state, props) => {
  return {
    expense: state.expenses.find(
      (expense) => expense.id === props.match.params.id
    ),
  };
};

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);
