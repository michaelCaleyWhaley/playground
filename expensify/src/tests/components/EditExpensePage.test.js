import React from "react";
import { shallow } from "enzyme";
import { EditExpensePage } from "../../components/EditExpensePage";
import expenses from "../fixtures/expenses";

describe("EditExpensePage", () => {
  let wrapper;
  let editExpense;
  let removeExpense;
  let history;
  let expense;

  beforeEach(() => {
    editExpense = jest.fn();
    removeExpense = jest.fn();
    history = { push: jest.fn() };
    expense = expenses[2];

    wrapper = shallow(
      <EditExpensePage
        editExpense={editExpense}
        removeExpense={removeExpense}
        expense={expense}
        history={history}
      />
    );
  });
  it("should render EditExpensePage", () => {
    expect(wrapper).toMatchSnapshot();
  });

  describe("when the expense form is submitted", () => {
    it("should call the editExpense prop", () => {
      expect(editExpense).not.toHaveBeenCalled();
      wrapper.find("ExpenseForm").prop("onSubmit")(expense);
      expect(editExpense).toHaveBeenCalledWith(expense.id, expense);
    });
  });

  describe("when the expense is removed", () => {
    it("should call the removeExpense props", () => {
      expect(removeExpense).not.toHaveBeenCalled();
      wrapper.find("button").simulate("click");
      expect(removeExpense).toHaveBeenCalledWith(expense.id);
    });
  });
});
