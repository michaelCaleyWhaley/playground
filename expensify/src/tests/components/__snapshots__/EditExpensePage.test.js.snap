// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditExpensePage should render EditExpensePage 1`] = `
<div>
  <ExpenseForm
    expense={
      Object {
        "amount": 4500,
        "createdAt": 864000000,
        "description": "Credit card",
        "id": "3",
        "note": "",
      }
    }
    onSubmit={[Function]}
  />
  <button
    onClick={[Function]}
  >
    Remove
  </button>
</div>
`;
